expr/func_highestlowest.go

```diff
@@ -0,0 +1,73 @@
+package expr
+
+import (
+	"sort"
+
+	"github.com/grafana/metrictank/consolidation"
+
+	"github.com/grafana/metrictank/api/models"
+)
+
+type FuncHighestLowest struct {
+	in      GraphiteFunc
+	n       int64
+	fn      string
+	highest bool
+}
+
+func NewHighestLowestConstructor(fn string, highest bool) func() GraphiteFunc {
+	return func() GraphiteFunc {
+		return &FuncHighestLowest{fn: fn, highest: highest}
+	}
+}
+
+func (s *FuncHighestLowest) Signature() ([]Arg, []Arg) {
+	if s.fn != "" {
+		return []Arg{
+			ArgSeriesList{val: &s.in},
+			ArgInt{key: "n", val: &s.n},
+		}, []Arg{ArgSeriesList{}}
+	}
+	return []Arg{
+		ArgSeriesList{val: &s.in},
+		ArgInt{key: "n", val: &s.n},
+		ArgString{key: "func", val: &s.fn, validator: []Validator{IsConsolFunc}},
+	}, []Arg{ArgSeriesList{}}
+}
+
+func (s *FuncHighestLowest) Context(context Context) Context {
+	return context
+}
+
+func (s *FuncHighestLowest) Exec(cache map[Req][]models.Series) ([]models.Series, error) {
+	series, err := s.in.Exec(cache)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(series) == 0 {
+		return series, nil
+	}
+
+	consolidationFunc := consolidation.GetAggFunc(consolidation.FromConsolidateBy(s.fn))
+
+	consolidationVals := make([]float64, len(series))
+
+	for i, serie := range series {
+		consolidationVals[i] = consolidationFunc(serie.Datapoints)
+	}
+
+	seriesLess := func(i, j int) bool {
+		if s.highest {
+			return consolidationVals[i] > consolidationVals[j]
```

Comment:
I don't think this will work, since `consolidationVals` won't change in step with `series` after the first swap in the algorithm, you'll start comparing against arbitrary other data.